src/gemdat/transitions.py

```diff
@@ -405,3 +406,197 @@ def _calculate_occupancy(atom_sites: np.ndarray) -> dict[int, int]:
+    """Wraps coordinates in periodic boundaries.
+
+    Parameters
+    ----------
+    coord: np.ndarray
+        Coordinates of one or multiple sites
+    size: tuple
+        Boudnary size
+
+    Returns:
+    -------
+    wrapped_coord: tuple
+        Coordinates inside the PBC
+    """
+    wrapped_coord = tuple(coord % size)
+    return wrapped_coord
+
+
+def free_energy_graph(F: np.ndarray,
+                      max_energy_threshold=1e20,
+                      diagonal=True) -> nx.Graph:
+    """Compute the graph of the free energy for networkx functions.
+
+    Parameters
+    ----------
+    F : np.ndarray
+        Free energy on the 3d grid
+    max_energy_threshold : float, optional
+        Maximum energy threshold for the path to be considered valid
+    diagonal : bool
+        If True, allows diagonal grid moves
+
+    Returns
+    -------
+    G : nx.Graph
+        Graph of free energy
+    """
+
+    # Define possible movements in 3D space
+    if not diagonal:
+        movements = [(1, 0, 0), (-1, 0, 0), (0, 1, 0), (0, -1, 0), (0, 0, 1),
+                     (0, 0, -1)]
+    else:
+        movements = [(1, 0, 0), (-1, 0, 0), (0, 1, 0), (0, -1, 0), (0, 0, 1),
+                     (0, 0, -1),
+                     (1, 1, 0), (-1, -1, 0), (1, -1, 0), (-1, 1, 0), (1, 0, 1),
+                     (-1, 0, -1), (1, 0, -1), (-1, 0, 1), (0, 1, 1),
+                     (0, -1, -1), (0, 1, -1), (0, -1, 1), (1, 1, 1),
+                     (-1, -1, -1), (1, -1, -1), (-1, 1, 1)]
+
+    G = nx.Graph()
+    for i in range(F.shape[0]):
+        for j in range(F.shape[1]):
+            for k in range(F.shape[2]):
+                if F[i, j, k] >= 0 and F[i, j, k] < max_energy_threshold:
+                    G.add_node((i, j, k), energy=F[i, j, k])
+    for i, j, k in G.nodes:
+        for move in movements:
+            neighbor = _wrap_pbc(np.array((i, j, k)) + np.array(move), F.shape)
+            if neighbor in G.nodes:
+                G.add_edge((i, j, k), neighbor, weight=F[neighbor])
+    return G
+
+
+def optimal_path(
+    F_graph: nx.Graph,
+    start: tuple,
+    end: tuple,
+) -> typing.Union[list, list]:
+    """Calculate the shortest cost-effective path from start to end using
+    Networkx library.
+
+    Parameters
+    ----------
+    F_graph : nx.Graph
+        Graph of the free energy
+    start : tuple
+        Coordinates of the starting point
+    end: tuple
+        Coordinates of the ending point
+
+    Returns
+    -------
+    path: list[tuple]
+        List of coordinates of the path
+    path_energy: list[float]
+        Energy along the path
+    """
+
+    optimal_path = nx.shortest_path(F_graph,
+                                    source=start,
+                                    target=end,
+                                    weight='weight')
+    path_energy = [F_graph.nodes[node]['energy'] for node in optimal_path]
+
+    return [optimal_path, path_energy]
```

Comment:
Does this need its own function?

Reply:
yes because it is also useful for other tasks, like issue #92

Reply:
The issue is that this function returns a tuple, which means this function does 2 things, which means I must deal with it in my code if I'm only interested in the path. This is poor separation of concerns. It's not obvious to me why it should return the energy as well. If you must return multiple values, it's usually easier in the long term to return an object (see other comment as well), and hide the energy list behind a method or property. Or split in two functions.

Personally, since this function is so trivial, I think it's easier to write it out in the function calling it.